deep_qa/training/train_utils.py

```diff
@@ -0,0 +1,113 @@
+from typing import List, Tuple
+import tensorflow
+
+
+def pin_variable_device_scope(device, variable_device="/cpu:0"):
+    """
+    Tensorflow device scopes can take functions which give a device
+    for a given op in the graph. Here, we use the device that is
+    passed to the scope *unless* the operation which is being created
+    in the graph is a Variable creation op; in this case, we place it
+    on the cpu.
+    """
+    def _assign(graph_op):
+        node_def = graph_op if isinstance(graph_op, tensorflow.NodeDef) else graph_op.node_def
+        if node_def.op in ['Variable', 'VariableV2']:
+            return variable_device
+        else:
+            return device
+    return _assign
+
+
+def average_gradients(tower_gradients: List[List[Tuple]]):
+    """
+    Given a list of (gradient, variable) pairs from the result of
+    a gradient calculation from multiple GPUs, calculate their
+    average.
+    """
+    # Make a map from variables -> [gradients that are not none].
+    gradient_map = {}
```

Comment:
`gradient_map = defaultdict(list)` lets you remove the `if variable not in gradient_map` check below.